File: t/PACSFramework/VB/DicomHighlevelPatientUpdaterDemo/OptionsDialog.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System.IO
Imports System.Net
Imports System.Text.RegularExpressions
Imports System.Runtime.InteropServices
Imports Microsoft.WindowsAPICodePack.Dialogs
Imports Leadtools.Dicom.Scu
Imports DicomDemo.DicomDemo.Dicom
Imports DicomDemo.DicomDemo.Utils

Namespace DicomDemo
   Partial Class OptionsDialog : Inherits Form
      Private _ServerAE As String

      Public Property ServerAE() As String
         Get
            Return _ServerAE
         End Get
         Set(ByVal value As String)
            _ServerAE = value
         End Set
      End Property

      Private _ServerIP As String

      Public Property ServerIP() As String
         Get
            Return _ServerIP
         End Get
         Set(ByVal value As String)
            _ServerIP = value
         End Set
      End Property

      Private _ServerPort As Integer

      Public Property ServerPort() As Integer
         Get
            Return _ServerPort
         End Get
         Set(ByVal value As Integer)
            _ServerPort = value
         End Set
      End Property

      Private _ClientAE As String

      Public Property ClientAE() As String
         Get
            Return _ClientAE
         End Get
         Set(ByVal value As String)
            _ClientAE = value
         End Set
      End Property

      Private _Timeout As Integer

      Public Property Timeout() As Integer
         Get
            Return _Timeout
         End Get
         Set(ByVal value As Integer)
            _Timeout = value
         End Set
      End Property

      Public Sub New()
         InitializeComponent()
      End Sub

      Private Function CheckInteger(ByVal tb As TextBox, ByVal lb As Label, ByVal showError As Boolean) As Boolean
         Try
            Convert.ToInt32(tb.Text)
            Return True
         Catch e1 As Exception
            If showError Then
               If tb.Text.Trim() = String.Empty Then
                  MessageBox.Show("Invalid " & lb.Text, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
               End If
               tb.SelectAll()
               tb.Focus()
               DialogResult = DialogResult.None
            End If
            Return False
         End Try
      End Function

      Private Function CheckIP(ByVal tb As TextBox, ByVal lb As Label, ByVal showError As Boolean) As Boolean
         Try
            System.Net.IPAddress.Parse(tb.Text)
            Return True
         Catch e1 As Exception
            If showError Then
               If tb.Text.Trim() = String.Empty Then
                  MessageBox.Show("Invalid " & lb.Text, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
               End If
               tb.SelectAll()
               tb.Focus()
               DialogResult = DialogResult.None
            End If
            Return False
         End Try
      End Function

      Public Function IsValidIP(ByVal addr As String) As Boolean
         'create our match pattern
         Dim pattern As String = "^([1-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])(\.([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])){3}$"
         'create our Regular Expression object
         Dim check As Regex = New Regex(pattern)

         'boolean variable to hold the status
         Dim valid As Boolean = False
         'check to make sure an ip address was provided
         If addr = "" Then
            'no address provided so return false
            valid = False
         Else
            'address provided so use the IsMatch Method
            'of the Regular Expression object
            valid = check.IsMatch(addr, 0)
         End If
         'return the results
         Return valid
      End Function

      Private Sub OptionsDialog_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
         _textBoxServerAE.Text = ServerAE
         _textBoxServerIP.Text = ServerIP
         _textBoxServerPort.Text = ServerPort.ToString()
         _textBoxClientAE.Text = ClientAE
         _textBoxTimeout.Text = Timeout.ToString()

         AddHandler Application.Idle, AddressOf Application_Idle
      End Sub

      Private Sub Application_Idle(ByVal sender As Object, ByVal e As EventArgs)
         If String.IsNullOrEmpty(_textBoxServerIP.Text) OrElse String.IsNullOrEmpty(_textBoxServerAE.Text) OrElse String.IsNullOrEmpty(_textBoxServerPort.Text) Then
            VerifyButton.Enabled = False
         Else
            VerifyButton.Enabled = IsValidIP(_textBoxServerIP.Text) AndAlso CheckInteger(_textBoxServerPort, Nothing, False)
         End If
      End Sub

      Private Sub _textBoxServerIP_KeyPress(ByVal sender As Object, ByVal e As KeyPressEventArgs) Handles _textBoxServerIP.KeyPress
         Dim bValid As Boolean = Char.IsNumber(e.KeyChar) OrElse (e.KeyChar = "."c) OrElse Char.IsControl(e.KeyChar)

         e.Handled = Not bValid
      End Sub

      Private Sub _textBoxServerPort_KeyPress(ByVal sender As Object, ByVal e As KeyPressEventArgs) Handles _textBoxServerPort.KeyPress
         If Not (Char.IsDigit(e.KeyChar) OrElse Char.IsControl(e.KeyChar)) Then
            e.Handled = True
         End If
      End Sub

      Private Sub VerifyButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles VerifyButton.Click
         Dim scp As DicomScp = New DicomScp()
         Dim scu As PatientUpdaterConnection = New PatientUpdaterConnection()

         scp.AETitle = _textBoxServerAE.Text
         scp.PeerAddress = IPAddress.Parse(_textBoxServerIP.Text)
         scp.Port = Convert.ToInt32(_textBoxServerPort.Text)
         scu.AETitle = _textBoxClientAE.Text
         Try
            Dim ret As Boolean = scu.Verify(scp)
            Dim scpInfo As String = scp.PeerAddress.ToString() & " (" & scp.Port.ToString() & ")"

            If (Not scu.Rejected) Then
               If ret Then
                  TaskDialogHelper.ShowMessageBox(Me, "Successfull Verification", "Verified " & scp.AETitle, String.Empty, scpInfo, TaskDialogStandardButtons.Ok, TaskDialogStandardIcon.Information, Nothing)
               Else
                  TaskDialogHelper.ShowMessageBox(Me, "Failed Verification", "Unable to verify " & scp.AETitle, String.Empty, scpInfo, TaskDialogStandardButtons.Ok, TaskDialogStandardIcon.Error, Nothing)
               End If
            Else
               TaskDialogHelper.ShowMessageBox(Me, "Failed Verification", "Unable to verify " & scp.AETitle, scu.Reason, scpInfo, TaskDialogStandardButtons.Ok, TaskDialogStandardIcon.Error, Nothing)
            End If
         Catch ex As Exception
            ApplicationUtils.ShowException(Me, ex)
         End Try
      End Sub

      Private Sub OptionsDialog_FormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs) Handles Me.FormClosing
         If (Not CheckIP(_textBoxServerIP, _labelServerIP, True)) Then
            e.Cancel = True
            Return
         End If

         If (Not CheckInteger(_textBoxServerPort, _labelServerPort, True)) Then
            e.Cancel = True
            Return
         End If

         If (Not CheckInteger(_textBoxTimeout, _labelTimeout, True)) Then
            e.Cancel = True
            Return
         End If

         ServerAE = _textBoxServerAE.Text
         ServerIP = _textBoxServerIP.Text
         ServerPort = Convert.ToInt32(_textBoxServerPort.Text)
         ClientAE = _textBoxClientAE.Text
         Timeout = Convert.ToInt32(_textBoxTimeout.Text)
      End Sub
   End Class
End Namespace

File: t/PACSFramework/CS/DicomHighlevelPatientUpdaterDemo/OptionsDialog.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.IO;
using Leadtools.Dicom.Scu;
using System.Net;
using System.Text.RegularExpressions;
using DicomDemo.Utils;
using System.Runtime.InteropServices;
using Microsoft.WindowsAPICodePack.Dialogs;
using DicomDemo.Dicom;

namespace DicomDemo
{
    public partial class OptionsDialog : Form
    {                        
        private string _ServerAE;

        public string ServerAE
        {
            get { return _ServerAE; }
            set { _ServerAE = value; }
        }

        private string _ServerIP;

        public string ServerIP
        {
            get { return _ServerIP; }
            set { _ServerIP = value; }
        }

        private int _ServerPort = 104;

        public int ServerPort
        {
            get { return _ServerPort; }
            set { _ServerPort = value; }
        }

        private string _ClientAE;

        public string ClientAE
        {
            get { return _ClientAE; }
            set { _ClientAE = value; }
        }

        private int _Timeout = 30;

        public int Timeout
        {
            get { return _Timeout; }
            set { _Timeout = value; }
        }

        public OptionsDialog()
        {
            InitializeComponent();
        }        

        private bool CheckInteger(TextBox tb, Label lb, bool showError)
        {
            try
            {
                Convert.ToInt32(tb.Text);
                return true;
            }
            catch (Exception)
            {
                if (showError)
                {
                    if (tb.Text.Trim() == string.Empty)
                        MessageBox.Show("Invalid " + lb.Text, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    tb.SelectAll();
                    tb.Focus();
                    DialogResult = DialogResult.None;
                }
                return false;
            }
        }        

        private bool CheckIP(TextBox tb, Label lb, bool showError)
        {
            try
            {
               if(tb.Text.Contains(','))
                  System.Net.IPAddress.Parse(tb.Text);
                return true;
            }
            catch (Exception)
            {
                if (showError)
                {
                    if (tb.Text.Trim() == string.Empty)
                        MessageBox.Show("Invalid " + lb.Text, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    tb.SelectAll();
                    tb.Focus();
                    DialogResult = DialogResult.None;
                }
                return false;
            }
        }

        public bool IsValidIP(string addr)
        {
            //create our match pattern
            string pattern = @"^([1-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])(\.([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])){3}$";
            //create our Regular Expression object
            Regex check = new Regex(pattern);

            if (!addr.Contains('.'))
               return true;

            //boolean variable to hold the status
            bool valid = false;
            //check to make sure an ip address was provided
            if (addr == "")
            {
                //no address provided so return false
                valid = false;
            }
            else
            {
                //address provided so use the IsMatch Method
                //of the Regular Expression object
                valid = check.IsMatch(addr, 0);
            }
            //return the results
            return valid;
        }                               
       
        private void OptionsDialog_Load(object sender, EventArgs e)
        {
            _textBoxServerAE.Text = ServerAE;
            _textBoxServerIP.Text = ServerIP;
            _textBoxServerPort.Text = ServerPort.ToString();
            _textBoxClientAE.Text = ClientAE;            
            _textBoxTimeout.Text = Timeout.ToString();               

            Application.Idle += new EventHandler(Application_Idle);
        }

        void Application_Idle(object sender, EventArgs e)
        {
            if(string.IsNullOrEmpty(_textBoxServerIP.Text) || string.IsNullOrEmpty(_textBoxServerAE.Text) ||
               string.IsNullOrEmpty(_textBoxServerPort.Text))
            {
                VerifyButton.Enabled = false;
            }
            else
            {
                VerifyButton.Enabled = IsValidIP(_textBoxServerIP.Text) && CheckInteger(_textBoxServerPort,null,false);                                       
            }
        }

        private void _textBoxServerIP_KeyPress(object sender, KeyPressEventArgs e)
        {
            //bool bValid = Char.IsNumber(e.KeyChar) || (e.KeyChar == '.') || Char.IsControl(e.KeyChar);

            //e.Handled = !bValid;
        }

        private void _textBoxServerPort_KeyPress(object sender, KeyPressEventArgs e)
        {
            if (!(Char.IsDigit(e.KeyChar) || Char.IsControl(e.KeyChar)))
            {
                e.Handled = true;
            }
        }

        private void VerifyButton_Click(object sender, EventArgs e)
        {
            DicomScp scp = new DicomScp();
            PatientUpdaterConnection scu = new PatientUpdaterConnection();

            scp.AETitle = _textBoxServerAE.Text;
            scp.PeerAddress = ResolveIPAddress(_textBoxServerIP.Text);
            scp.Port = Convert.ToInt32(_textBoxServerPort.Text);
            scu.AETitle = _textBoxClientAE.Text;            
            try
            {
                bool ret = scu.Verify(scp);
                string scpInfo = scp.PeerAddress.ToString() + " (" + scp.Port.ToString() + ")";

                if (!scu.Rejected)
                {
                    if (ret)
                    {
                        TaskDialogHelper.ShowMessageBox(this, "Successful Verification", "Verified " + scp.AETitle, string.Empty,
                                                        scpInfo, TaskDialogStandardButtons.Ok, TaskDialogStandardIcon.Information, null);
                    }
                    else
                    {
                        TaskDialogHelper.ShowMessageBox(this, "Failed Verification", "Unable to verify " + scp.AETitle, string.Empty,
                                                        scpInfo, TaskDialogStandardButtons.Ok, TaskDialogStandardIcon.Error, null);
                    }
                }
                else
                {                    
                    TaskDialogHelper.ShowMessageBox(this, "Failed Verification", "Unable to verify " + scp.AETitle, scu.Reason,
                                                          scpInfo, TaskDialogStandardButtons.Ok, TaskDialogStandardIcon.Error, null);
                }
            }
            catch (Exception ex)
            {
                ApplicationUtils.ShowException(this, ex);
            }
        }

        public static System.Net.IPAddress ResolveIPAddress(string hostNameOrAddress)
        {
           IPAddress[] addresses;
           addresses = Dns.GetHostAddresses(hostNameOrAddress.Trim());
           if (addresses == null || addresses.Length == 0)
           {
              throw new ArgumentException("Invalid hostNameOrAddress parameter.");
           }
           else
           {
              IPAddress address = (from a in addresses.OfType<IPAddress>()
                                   where a.AddressFamily == System.Net.Sockets.AddressFamily.InterNetwork
                                   select a).FirstOrDefault();

              if (address == null)
              {
                 address = (from a in addresses.OfType<IPAddress>()
                            where a.AddressFamily == System.Net.Sockets.AddressFamily.InterNetworkV6
                            select a).FirstOrDefault();
              }

              if (address != null)
                 return address;

              throw new ArgumentException("Could not resolve a valid host Address. Address must conform to IPv4 or IPv6.");
           }
        }

        private void OptionsDialog_FormClosing(object sender, FormClosingEventArgs e)
        {
           if (DialogResult == DialogResult.OK)
           {
              if (!CheckIP(_textBoxServerIP, _labelServerIP, true))
              {
                 e.Cancel = true;
                 return;
              }

              if (!CheckInteger(_textBoxServerPort, _labelServerPort, true))
              {
                 e.Cancel = true;
                 return;
              }

              if (!CheckInteger(_textBoxTimeout, _labelTimeout, true))
              {
                 e.Cancel = true;
                 return;
              }

              ServerAE = _textBoxServerAE.Text;
              ServerIP = _textBoxServerIP.Text;
              ServerPort = Convert.ToInt32(_textBoxServerPort.Text);
              ClientAE = _textBoxClientAE.Text;
              Timeout = Convert.ToInt32(_textBoxTimeout.Text);
           }
        }

        private void OptionsDialog_Shown(object sender, EventArgs e)
        {
           _textBoxServerAE.Focus();
        }                
    }
}
